File: iOS/Objective-C/ObjectiveC Study/ObjectiveC Study/Class/Inheritance.m
//
//  Inheritance.m
//  ObjectiveC Study
//
//  Created by Harry Xiong on 2024/6/21.
//

#import <Foundation/Foundation.h>
#import "Inheritance.h"

// 1. 重写：子类重新实现父类中的某个方法，覆盖父类以前的做法
// 2. 父类必须声明在子类的前面
// 3. 子类不能拥有和父类相同的成员变量
// 4. 调用某个对象的方法时，优先去当前类中找，如果找找不到，到父类去找

@implementation Animal

- (void)setAge:(int)age
{
    self->_age = age;
}

- (int)age
{
    return self->_age;
}

- (void)setWeight:(double)weight
{
    self->_weight = weight;
}

- (double)weight
{
    return self->_weight;
}

@end

@implementation Fish

@end

@implementation Cat

@end

void testInheritance(void) {
    Fish *f = [Fish new];
    [f setAge:10];
    NSLog(@"fish age=%d", [f age]);
    
    Cat *c = [Cat new];
    [c setWeight:12.0];
    NSLog(@"cat weight=%.2f", [c weight]);
}

File: iOS/Swift/Swift Study/Swift Study/Class/Inheritance.swift
//
//  Inheritance.swift
//  Swift Study
//
//  Created by Harry Xiong on 2024/6/24.
//

import Foundation

class Animal {
    private var age: Int;
    private var weight: Double;
    
    var computedAge: Int  {
        get {
            return self.age
        }
        set {
            self.age = newValue
        }
    }
    
    var computedWeight: Double {
        get {
            return self.weight
        }
        set {
            self.weight = newValue
        }
    }
    
    init(age: Int, weight: Double) {
        self.age = age
        self.weight = weight
    }
    
    func description() {
        print("Age: \(self.age), Weight: \(self.weight).")
    }
}

class Fish: Animal {
    
    let fishCategory: String;
    
    init(age: Int, weight: Double, fishCategroy: String) {
        self.fishCategory = fishCategroy
        super.init(age: age, weight: weight);
    }

}

class Cat: Animal {
    
    let catCategory: String;
    
    init(age: Int, weight: Double, catCategory: String) {
        self.catCategory = catCategory
        super.init(age: age, weight: weight);
    }
}

func testInheritance() {
    let fish = Fish(age: 20, weight: 200, fishCategroy: "Fish A")
    let cat = Cat(age: 10, weight: 100, catCategory: "Cat A")
    
    print(fish.computedAge, cat.computedAge)
    
    fish.description()
    cat.description()
}

